rust_src/src/coding.rs

```diff
@@ -0,0 +1,38 @@
+//! Coding system handler.
+
+use libc::c_int;
+
+use remacs_macros::lisp_fn;
+
+use crate::{
+    hashtable::HashLookupResult::{Found, Missing},
+    lisp::defsubr,
+    lisp::LispObject,
+    remacs_sys::{Fget, Qcoding_system_define_form, Vcoding_system_hash_table},
+};
+
+/// Return the ID of CODING_SYSTEM_SYMBOL.
+#[no_mangle]
+pub extern "C" fn CODING_SYSTEM_ID(object: LispObject) -> c_int {
+    let h_ref = unsafe { Vcoding_system_hash_table }.as_hash_table_or_error();
+    match h_ref.lookup(object) {
+        Found(idx) => idx as c_int,
+        Missing(_) => -1,
+    }
+}
+
+/// Return t if OBJECT is nil or a coding-system.
+/// See the documentation of `define-coding-system' for information
+/// about coding-system objects.
+#[lisp_fn]
+pub fn coding_system_p(object: LispObject) -> bool {
+    if object.is_nil() || CODING_SYSTEM_ID(object) >= 0 {
```

Comment:
```suggestion
    (object.is_nil() || CODING_SYSTEM_ID(object) >= 0) && !(!object.is_symbol || unsafe { Fget(object, Qcoding_system_define_form) }.is_nil())
```

Reply:
The C code often returns `Qt` or `Qnil`, but that's because C functions need to return `Lisp_Object` every time. `lisp_fn` takes care of the conversions in Rust, so boolean logic can be simplified to avoid returning plain `true` or `false`.

Reply:
If I'm reading the logic right, there are three cases where this function returns `t`: 1) the object is null; 2) the object's coding system id is at least zero; and 3) the object is a symbol with a nonnull `coding-system-define-form` property.

Reply:
If this path is taken, apply deMorgan's and simplify.

```
(object.is_nil() || coding_system_id(object)) ||
(object.is_symbol() && !unsafe { Fget(object, Qcoding_system_define_form) }.is_nil())
```

Reply:
Yeah, I realized that right after I posted it. The C code seems to violate it a lot.